architecture/additional_concepts/authentication.adoc

```diff
@@ -345,6 +345,30 @@ desired behavior:
 "serviceaccounts.openshift.io/oauth-redirecturi.second": "https://other.com"
 ----
 
+[[api-events-oauth-clients]]
+==== API Events for OAuth
+
+In some cases the API server returns an unexpected internal error message 
+that are hard to debug without direct access to the API master log. 
+These errors must be masked for security reasons as they are OAuth errors that 
+organizations might not want to let non-administrator users to see. 
+
+In lot of cases these errors can be corrected by non-administrator users, 
+however without access to the master log, the non-admin users cannot get the
+necessary information.
+
+Events associated with service accounts can be found in the `oc get events` output 
+to allow non-administrator users to debug these issues without access to the API master logs.
+
+In the event of an OAuth `unexpected internal error`, use the `oc get events` command and search for `OAuth*` entries in the *Reason* column.
+
+----
+oc get events | grep OAuth*
+LASTSEEN   FIRSTSEEN   COUNT     NAME      KIND             SUBOBJECT   TYPE      REASON                   SOURCE                                   MESSAGE
+1m         1m          1         foobar    ServiceAccount               Normal    OAuthAllRedirectURIs     {service-account-oauth-client-getter }   Has the following redirectURIs: [https://foo]
+1m         1m          1         foobar    ServiceAccount               Warning   OAuthAnnotationSkipped   {service-account-oauth-client-getter }   Annotation value is not a valid URL: serviceaccounts.openshift.io/oauth-redirecturi.foo=&*^&*^&*^&*$%$%^$%^$
+----
+
```

Comment:
There are a few things with these paragraphs that we might want to say differently, I would suggest the following text. The command examples are also updated per the current code.

In some cases the API server will return an "unexpected condition" error message that is difficult to debug without direct access to the API master log. The underlying reason for the error is purposely obscured in order to avoid providing an unauthenticated user with information about the server's state.

A subset of these errors may be related to a service account OAuth configuration issue, and will be captured in events that can be viewed by non-administrator users. When encountering an "unexpected condition" server error during OAuth, running `oc get events` will display these events under the `ServiceAccount` kind.

The following example warns of a service account that is missing a proper OAuth redirect URI:
````
$ oc get events | grep ServiceAccount
1m         1m          1         proxy                    ServiceAccount                                  Warning   NoSAOAuthRedirectURIs   service-account-oauth-client-getter   system:serviceaccount:myproject:proxy has no redirectURIs; set serviceaccounts.openshift.io/oauth-redirecturi.<some-value>=<redirect> or create a dynamic URI using serviceaccounts.openshift.io/oauth-redirectreference.<some-value>=<reference>
````

Running `oc describe sa/<service-account-name>` will also display any OAuth events associated with the given service account name.
````
$ oc describe sa/proxy | grep -A5 Events
Events:
  FirstSeen     LastSeen        Count   From                                    SubObjectPath   Type            Reason                  Message
  ---------     --------        -----   ----                                    -------------   --------        ------                  -------
  3m            3m              1       service-account-oauth-client-getter                     Warning         NoSAOAuthRedirectURIs   system:serviceaccount:myproject:proxy has no redirectURIs; set serviceaccounts.openshift.io/oauth-redirecturi.<some-value>=<redirect> or create a dynamic URI using serviceaccounts.openshift.io/oauth-redirectreference.<some-value>=<reference>
````